### PR TITLE
Process: support execvpe and read error from child

### LINKED ISSUE #3: [Bug] inconsistent behavior between *nix and Windows
This issue is related to #2 , which has some TODOs.

### PR DESCRIPTION
**Pull request by imkiva**
This fixes #3

### CODE REVIEW

File: src/process_win32.cpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+ * Website: https://covariant.cn/
+ * Github:  https://github.com/covariant-institute/
+ */
+#include <mozart++/core>
+#ifdef MOZART_PLATFORM_WIN32
+
+#include <mozart++/process>
+#include <Windows.h>
+
+namespace mmpp_impl {
```
Comment (mikecovlee) [inline src/process_win32.cpp]: typo here: mmpp_impl -> mpp_impl

File: src/process_unix.cpp [hunk 2]
```diff
@@ -0,0 +1,550 @@
+/**
+ * Mozart++ Template Library
+ * Licensed under MIT License
+ * Copyright (c) 2020 Covariant Institute
+ * Website: https://covariant.cn/
+ * Github:  https://github.com/covariant-institute/
+ */
+#include <mozart++/core>
+
+#ifndef MOZART_PLATFORM_WIN32
```
Comment (mikecovlee) [inline src/process_unix.cpp]: Please use `MOZART_PLATFORM_UNIX` here

File: src/process_unix.cpp [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+#include <cstdlib>
+#include <cstring>
+#include <unistd.h>
+#include <cctype>
+#include <climits>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <csignal>
+
+#ifdef __APPLE__
```
Comment (mikecovlee) [inline src/process_unix.cpp]: Please replace all of the `__APPLE__` with `MOZART_PLATFORM_DARWIN`

File: src/process_unix.cpp [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+#define FD_DIR "/dev/fd"
+#define dirent64 dirent
+#define readdir64 readdir
+#else
+#define FD_DIR "/proc/self/fd"
+#endif
+
+#ifdef __APPLE__
```
Comment (mikecovlee) [inline src/process_unix.cpp]: same as above

File: src/process_unix.cpp [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+                        // TODO: should we report this as an error?
+                        continue;
+                    }
+                }
+            }
+
+            // change cwd
+            if (chdir(startup._cwd.c_str()) != 0) {
+                // do not use mpp::throw_ex
+                throw mpp::runtime_error();
```
Comment (mikecovlee) [inline src/process_unix.cpp]: Consider using `mpp::throw_ex` because it can handle noexcept situations.

File: src/process_unix.cpp [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+            // there are two cases of this situation depending on signal set
+            struct sigaction sa{};
+            if (sigaction(SIGCHLD, nullptr, &sa) != 0) {
+                // only happens when kernel bug
+                mpp::throw_ex<mpp::runtime_error>("should not reach here");
+            }
+
+#if defined(__APPLE__)
+            void *handler = reinterpret_cast<void *>(sa.__sigaction_u.__sa_handler);
+#elif defined(__linux__)
```
Comment (mikecovlee) [inline src/process_unix.cpp]: use `MOZART_PLATFORM_LINUX` instead.

File: src/process_unix.cpp [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+
+            // waitpid() cannot find the child process identified by pid,
+            // there are two cases of this situation depending on signal set
+            struct sigaction sa{};
+            if (sigaction(SIGCHLD, nullptr, &sa) != 0) {
+                // only happens when kernel bug
+                mpp::throw_ex<mpp::runtime_error>("should not reach here");
+            }
+
+#if defined(__APPLE__)
```
Comment (mikecovlee) [inline src/process_unix.cpp]: use `MOZART_PLATFORM_DARWIN` instead.

**Review by mikecovlee — CHANGES_REQUESTED** [review]
Still lacking some details

File: src/process_win32.cpp (hunk 1 above)
Comment (imkiva) [reply src/process_win32.cpp]: Fixed

File: src/process_unix.cpp (hunk 5 above)
Comment (imkiva) [reply src/process_unix.cpp]: No. noexcept will be handle with `std::terminate`, which is out-of-control.
And I don't know what happens inside this mysterious C++ function.
Here the parent are waiting the child to perform some IO operations (e.g. close or write) on the fail pipe, BUT I am not sure if **ALL** `std::terminate` will do the cleanup job.

I using raw throw here is just to jump the control flow to the cleanup aera, which will ensure
the parent-child-communication pipe will be closed properly, thus ensuring the parent process will not get stuck in `read_fully`.

Comment (imkiva) [reply src/process_unix.cpp]: So the best way might be `goto`...?

Comment (imkiva) [reply src/process_unix.cpp]: Fuck me, solved.

**Review by mikecovlee — APPROVED** [review]
Go.

**Pull request merged**